dialog/fileicon.go

```diff
@@ -0,0 +1,127 @@
+package dialog
+
+import (
+	"bufio"
+	"image/color"
+	"mime"
+	"os"
+	"path/filepath"
+	"strings"
+	"unicode/utf8"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/canvas"
+	"fyne.io/fyne/theme"
+	"fyne.io/fyne/widget"
+)
+
+type fileIcon struct {
+	widget.BaseWidget
+
+	extension, mimeType, mimeSubType string
+	resource fyne.Resource
+}
+
+const ratioDown float64 = 0.45
+
+func (i *fileIcon) CreateRenderer() fyne.WidgetRenderer {
+	img := canvas.NewImageFromResource(i.resource)
+	extText := canvas.NewText(i.extension, theme.BackgroundColor())
+	extText.Alignment = fyne.TextAlignCenter
+	extText.TextSize = theme.TextSize()
+	return &fileIconRenderer{item: i,
+		img: img, ext: extText, objects: []fyne.CanvasObject{img, extText}}
+}
+
+func mimeTypeGet(path string) (ext, mimeType, mimeSubType string) {
+	ext = filepath.Ext(path[1:])
+	mimeTypeFull := mime.TypeByExtension(ext)
+	if mimeTypeFull == "" {
+		mimeTypeFull = "text/plain"
+		file, err := os.Open(path)
+		if err == nil {
+			defer file.Close()
+			scanner := bufio.NewScanner(file)
+			if scanner.Scan() && !utf8.Valid(scanner.Bytes()) {
+				mimeTypeFull = "application/octet-stream"
+			}
+		}
+	}
+
+	mimeTypeSplit := strings.Split(mimeTypeFull, "/")
+	mimeType = mimeTypeSplit[0]
+	mimeSubTypeSplit := strings.Split(mimeTypeSplit[1], ";")
```

Comment:
This will panic if we have not got a "/" in the string. perhaps we should be more safe?